tests/datastore_firestore/test_async_transaction.py

```diff
@@ -0,0 +1,128 @@
+from newrelic.api.background_task import background_task
+from testing_support.validators.validate_database_duration import (
+    validate_database_duration,
+)
+
+
+@pytest.fixture(autouse=True)
+def sample_data(collection):
+    for x in range(1, 4):
+        collection.add({"x": x}, "doc%d" % x)
+
+
+@pytest.fixture()
+def exercise_async_transaction_commit(async_client, async_collection):
+    async def _exercise_async_transaction_commit():
+        from google.cloud.firestore import async_transactional
+        
+        @async_transactional
+        async def _exercise(async_transaction):
+            # get a DocumentReference
+            with pytest.raises(TypeError):  # get is currently broken. It attempts to await an async_generator instead of consuming it.
+                [_ async for _ in async_transaction.get(async_collection.document("doc1"))]
+
+            # get a Query
+            with pytest.raises(TypeError):  # get is currently broken. It attempts to await an async_generator instead of consuming it.
+                async_query = async_collection.select("x").where(field_path="x", op_string=">", value=2)
+                assert len([_ async for _ in async_transaction.get(async_query)]) == 1
+
+            # get_all on a list of DocumentReferences
+            with pytest.raises(TypeError):  # get_all is currently broken. It attempts to await an async_generator instead of consuming it.
+                all_docs = async_transaction.get_all([async_collection.document("doc%d" % x) for x in range(1, 4)])
+                assert len([_ async for _ in all_docs]) == 3
+
+            # set and delete methods
+            async_transaction.set(async_collection.document("doc2"), {"x": 0})
+            async_transaction.delete(async_collection.document("doc3"))
+
+        await _exercise(async_client.transaction())
+        assert len([_ async for _ in async_collection.list_documents()]) == 2
+    return _exercise_async_transaction_commit
+
+
+@pytest.fixture()
+def exercise_async_transaction_rollback(async_client, async_collection):
+    async def _exercise_async_transaction_rollback():
+        from google.cloud.firestore import async_transactional
+        
+        @async_transactional
+        async def _exercise(async_transaction):
+            # set and delete methods
+            async_transaction.set(async_collection.document("doc2"), {"x": 99})
+            async_transaction.delete(async_collection.document("doc1"))
+            raise RuntimeError()
+
+        with pytest.raises(RuntimeError):
+            await _exercise(async_client.transaction())
+        assert len([_ async for _ in async_collection.list_documents()]) == 3
+    return _exercise_async_transaction_rollback
+
+
+def test_firestore_async_transaction_commit(loop, exercise_async_transaction_commit, async_collection):
+    _test_scoped_metrics = [
+        ("Datastore/operation/Firestore/commit", 1),
+        # ("Datastore/operation/Firestore/get_all", 2),
```

Comment:
Commented out code left in here

Reply:
Ah yep, leaving that there because it's what the tests SHOULD be doing but AsyncTransaction is broken. PR is open to Firestore to fix their library, we should see when this starts to work and can add this metric back in.